src/Microsoft.AspNet.Security/Infrastructure/SecurityHelper.cs

```diff
@@ -17,18 +17,20 @@ namespace Microsoft.AspNet.Security.Infrastructure
     public static class SecurityHelper
     {
         /// <summary>
-        /// Add an additional ClaimsIdentity to the ClaimsPrincipal
+        /// Add all ClaimsIdenities from an additional ClaimPrincipal to the ClaimsPrincipal
         /// </summary>
         /// <param name="identity"></param>
-        public static void AddUserIdentity([NotNull] HttpContext context, [NotNull] IIdentity identity)
+        public static void AddUserPrincipal([NotNull] HttpContext context, [NotNull] IPrincipal principal)
         {
-            var newClaimsPrincipal = new ClaimsPrincipal(identity);
+            var newClaimsPrincipal = new ClaimsPrincipal(principal);
```

Comment:
Does HttpContext.User have a default ClaimsPrincipal? This may be driving the whole check about existing principal. In the 90% case, a single AH will create a single ClaimsPrincipal. WsFed, OIDC follow this model.